File: code/packages/csharp/algol-parser/AlgolParser.cs
using System.Collections.Concurrent;
using System.Text;
using CodingAdventures.GrammarTools;
using CodingAdventures.Parser;

namespace CodingAdventures.AlgolParser;

/// <summary>
/// ALGOL parser backed by shared grammar definitions.
/// </summary>
public static class AlgolParser
{
    public static string DefaultVersion => CodingAdventures.AlgolLexer.AlgolLexer.DefaultVersion;
    public static IReadOnlyList<string> SupportedVersions => CodingAdventures.AlgolLexer.AlgolLexer.SupportedVersions;

    private static readonly ConcurrentDictionary<string, ParserGrammar> ParserGrammars = new();

    public static GrammarParser CreateAlgolParser() => CreateAlgolParser(DefaultVersion);

    public static GrammarParser CreateAlgolParser(string? version) => new(LoadParserGrammar(version));

    public static ASTNode ParseAlgol(string source) => ParseAlgol(source, DefaultVersion);

    public static ASTNode ParseAlgol(string source, string? version)
    {
        try
        {
            return CreateAlgolParser(version).Parse(CodingAdventures.AlgolLexer.AlgolLexer.TokenizeAlgol(source, version));
        }
        catch (GrammarParseError error)
        {
            throw new ArgumentException("ALGOL parse failed: " + error.Message, nameof(source), error);
        }
    }

    private static ParserGrammar LoadParserGrammar(string? version)
    {
        var validated = ValidateVersion(version);
        return ParserGrammars.GetOrAdd(validated, ParseParserGrammarResource);
    }

    private static string ValidateVersion(string? version)
    {
        if (string.IsNullOrWhiteSpace(version))
        {
            return DefaultVersion;
        }

        if (!SupportedVersions.Contains(version))
        {
            throw new ArgumentException("Unknown ALGOL version '" + version + "'. Valid values: " + string.Join(", ", SupportedVersions), nameof(version));
        }

        return version;
    }

    private static ParserGrammar ParseParserGrammarResource(string version)
    {
        try
        {
            return ParserGrammarParser.Parse(ReadResource(version + ".grammar"));
        }
        catch (ParserGrammarError error)
        {
            throw new InvalidOperationException("Failed to parse bundled ALGOL parser grammar for version " + version, error);
        }
    }

    private static string ReadResource(string resourceName)
    {
        var assembly = typeof(AlgolParser).Assembly;
        using var stream = assembly.GetManifestResourceStream(resourceName)
            ?? throw new InvalidOperationException("Missing bundled resource: " + resourceName);
        using var reader = new StreamReader(stream, Encoding.UTF8);
        return reader.ReadToEnd();
    }
}

File: code/packages/fsharp/algol-parser/AlgolParser.fs
namespace CodingAdventures.AlgolParser.FSharp

open System
open System.Collections.Concurrent
open System.IO
open System.Text
open CodingAdventures.GrammarTools.FSharp
open CodingAdventures.Parser.FSharp
open CodingAdventures.AlgolLexer.FSharp

/// ALGOL parser backed by shared grammar definitions.
type AlgolParser private () =
    static let parserGrammars = ConcurrentDictionary<string, ParserGrammar>()

    static member DefaultVersion = AlgolLexer.DefaultVersion
    static member SupportedVersions = AlgolLexer.SupportedVersions

    static member CreateAlgolParser() = AlgolParser.CreateAlgolParser(AlgolParser.DefaultVersion)
    static member CreateAlgolParser(version: string) = GrammarParser(AlgolParser.LoadParserGrammar(version))

    static member ParseAlgol(source: string) = AlgolParser.ParseAlgol(source, AlgolParser.DefaultVersion)
    static member ParseAlgol(source: string, version: string) =
        try
            AlgolParser.CreateAlgolParser(version).Parse(AlgolLexer.TokenizeAlgol(source, version))
        with
        | :? GrammarParseError as error ->
            raise (ArgumentException(sprintf "ALGOL parse failed: %s" error.Message, "source", error))

    static member private LoadParserGrammar(version: string) =
        let validated = AlgolParser.ValidateVersion(version)
        parserGrammars.GetOrAdd(validated, Func<string, ParserGrammar>(AlgolParser.ParseParserGrammarResource))

    static member private ValidateVersion(version: string) =
        if String.IsNullOrWhiteSpace(version) then
            AlgolParser.DefaultVersion
        elif AlgolParser.SupportedVersions |> Seq.exists ((=) version) then
            version
        else
            invalidArg "version" (sprintf "Unknown ALGOL version '%s'. Valid values: %s" version (String.Join(", ", AlgolParser.SupportedVersions)))

    static member private ParseParserGrammarResource(version: string) =
        try
            ParserGrammarParser.Parse(AlgolParser.ReadResource(version + ".grammar"))
        with
        | :? ParserGrammarError as error ->
            raise (InvalidOperationException(sprintf "Failed to parse bundled ALGOL parser grammar for version %s" version, error))

    static member private ReadResource(resourceName: string) =
        let assembly = typeof<AlgolParser>.Assembly
        use stream = assembly.GetManifestResourceStream(resourceName)
        if isNull stream then
            invalidOp (sprintf "Missing bundled resource: %s" resourceName)
        use reader = new StreamReader(stream, Encoding.UTF8)
        reader.ReadToEnd()
